### ISSUE TITLE
Remove cyclic dependency among cloud-api-adaptor and peerpod-ctrl

### ISSUE THREAD

**Issue by wainersm** [confidential-containers/cloud-api-adaptor#1122]
As explained by @yoheiueda in https://github.com/confidential-containers/cloud-api-adaptor/pull/1110#issuecomment-1603914981 currently we have a cyclic dependency among the cloud-api-adaptor and peerpod-ctl go modules. He also suggest the creating of a `cloud` module could break the cycle.

Usually cyclic dependency signals something is wrong with API design.

Cc @stevenhorsman

**Comment by Amulyam24** [comment]
Hi @bpradipt, I'll take this up.

**Comment by surajssd** [comment]
So will we have separate repositories called controllers?

**Comment by yoheiueda** [comment]
@surajssd My original intention was to create a new Go sub module under the same cloud-api-adaptor GitHub repo.

So, the list of the Go modules will look like this. Each module has  its own `go.mod` file.

* github.com/confidential-containers/cloud-api-adaptor
* github.com/confidential-containers/cloud-api-adaptor/cloud ← **New module**
* github.com/confidential-containers/cloud-api-adaptor/peerpod-ctrl
* github.com/confidential-containers/cloud-api-adaptor/peerpodconfig-ctrl
* github.com/confidential-containers/cloud-api-adaptor/volumes/csi-wrapper
* github.com/confidential-containers/cloud-api-adaptor/webhook

I think separating controller code into other repositories is another topic, but I also think it is an reasonable option to consider.

**Comment by bpradipt** [comment]
> So will we have separate repositories called controllers?
What will make release process easier ?

So if we have the following structure, will it help with release and ongoing maintenance ?

github.com/confidential-containers/cloud-api-adaptor
github.com/confidential-containers/cloud-api-adaptor/cloud ← New module
github.com/confidential-containers/cloud-api-adaptor/volumes/csi-wrapper
github.com/confidential-containers/cloud-api-adaptor/webhook

and 
github.com/confidential-containers/<controller-prj-name>/{peerpodconfig-ctrl, peerpod-ctrl}

Both peerpodconfig and peerpod controllers are independent of each other, so can be in one repo. But should we keep this in their own repo for ease of release management and maintenance ?

**Comment by yoheiueda** [comment]
@bpradipt I think it will work in the perspectives from Go modules

Please note that the repositories will still have mutual dependencies, since the first repo contains both `cloud-api-adaptor` and `cloud-api-adaptor/cloud` modules, and there are dependencies below.

peerpod-ctrl → cloud-api-adaptor/cloud 
cloud-api-adaptor → peerpod-ctrl 

This means that we will still need to be careful when we prepare release tags.

**Comment by bpradipt** [comment]
This dep will still create issues
```
peerpod-ctrl → cloud-api-adaptor/cloud
cloud-api-adaptor → peerpod-ctrl
```
Do you have any thoughts on how this cyclic dep between CAA and peerpod-ctr be avoided ?
Should we look at a extracting out the cloud module in a separate repo which then gets imported in CAA and peerpod-ctrl ?

**Comment by yoheiueda** [comment]
@bpradipt 

We can create a Git tag per Go modules in a single repository as described in https://github.com/confidential-containers/cloud-api-adaptor/issues/852 and here https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/Release-Process.md#tags-and-update-go-submodules

For example, we can create two tags `v0.8.0` and `cloud/v0.8.0`. The first one is used to specify the version of the Go pacakge `github.com/confidential-containers/cloud-api-adaptor`, and the second one is used to specify the version of `github.com/confidential-containers/cloud-api-adaptor/cloud`.

Each tag can refer to a different commit, so we can create Git tags and update `go.mod` in the following order.

1. cloud-api-adaptor/cloud
2. peerpod-ctrl
3. cloud-api-adaptor

Another possible solution is to move [the `peerpod-ctrl/api/v1alpha1` pacakge](https://github.com/confidential-containers/cloud-api-adaptor/tree/4111acc80cda3a332927a8b34c009d887c7cbae5/peerpod-ctrl/api/v1alpha1) from `peerpod-ctrl` to `cloud-api-adaptor`. This package is generated one, and referenced from both the `cloud-api-adaptor` binary and the controller binary.

**Comment by wainersm** [comment]
This issue come up over and over again every time we make a release, and that is the case for v0.8.0.

Let's see if I understood [yoheiueda](https://github.com/yoheiueda)'s proposal below and maybe we can reach on a consensus to finally fix it.

> @bpradipt
> 
> We can create a Git tag per Go modules in a single repository as described in #852 and here https://github.com/confidential-containers/cloud-api-adaptor/blob/main/docs/Release-Process.md#tags-and-update-go-submodules
> 
> For example, we can create two tags `v0.8.0` and `cloud/v0.8.0`. The first one is used to specify the version of the Go pacakge `github.com/confidential-containers/cloud-api-adaptor`, and the second one is used to specify the version of `github.com/confidential-containers/cloud-api-adaptor/cloud`.
> 
> Each tag can refer to a different commit, so we can create Git tags and update `go.mod` in the following order.
> 
>     1. cloud-api-adaptor/cloud
> 

Step 1. Create the `cloud/v0.8.0` tag

>     2. peerpod-ctrl
> 

Step 2. `peerpod-ctrl` -> `cloud-api-adaptor/cloud@cloud/v0.8.0`
```
cd peerpod-ctrl
go get github.com/confidential-containers/cloud-api-adaptor/cloud@cloud/v0.8.0
```

Open a PR and merge.

>     3. cloud-api-adaptor
>

Step 3. `cloud-api-adaptor` -> `cloud-api-adaptor/cloud@cloud/v0.8.0`
```
go get github.com/confidential-containers/cloud-api-adaptor/cloud@cloud/v0.8.0
```

Open a PR and merge.

step 4. Would be `cloud-api-adaptor` -> `peerpod-ctrl` IIUC. Here I got confused. What tag to use?

Step N. There is also the `volumes/csi-wrapper` -> `cloud-api-adaptor`

> 
> Another possible solution is to move [the `peerpod-ctrl/api/v1alpha1` pacakge](https://github.com/confidential-containers/cloud-api-adaptor/tree/4111acc80cda3a332927a8b34c009d887c7cbae5/peerpod-ctrl/api/v1alpha1) from `peerpod-ctrl` to `cloud-api-adaptor`. This package is generated one, and referenced from both the `cloud-api-adaptor` binary and the controller binary.

**Comment by stevenhorsman** [comment]
JT has suggested that we might be able to do some of the cloud module separation in #1508

**Comment by liudalibj** [comment]
This issue blocked #1706 I will work on this item and try to work out a fix.

**Comment by bpradipt** [comment]
@liudalibj can we move peerpod-ctrl peerpodconfig-ctrl to separate repo as it'll fix the cycling dep issues ?

**Comment by liudalibj** [comment]
> @liudalibj can we move peerpod-ctrl peerpodconfig-ctrl to separate repo as it'll fix the cycling dep issues ?

Move peerpod-ctrl peerpodconfig-ctrl to separate repo won't fix the cycling dep issue.

the root cause for the cycling dep issue:
1. cloud-api-adaptor -> peerpod-ctrl https://github.com/confidential-containers/cloud-api-adaptor/blob/main/pkg/adaptor/k8sops/peerpod.go, caa need `"github.com/confidential-containers/cloud-api-adaptor/peerpod-ctrl/api/v1alpha1"`

2. peerpod-ctrl ->  cloud-api-adaptor https://github.com/confidential-containers/cloud-api-adaptor/blob/main/peerpod-ctrl/controllers/peerpod_controller.go, peerpod-ctrl need `"github.com/confidential-containers/cloud-api-adaptor/pkg/adaptor/cloud"`

So the possible two fixes from @yoheiueda :
1. New a module: "github.com/confidential-containers/cloud-api-adaptor/cloud", 
Then the dep will looks like:
```
1. peerpod-ctrl → cloud
2. cloud-api-adaptor → peerpod-ctrl (→ cloud)
3. csi-wrapper → cloud-api-adaptor (→ peerpod-ctrl → cloud)
```
2.  move [the peerpod-ctrl/api/v1alpha1 pacakge](https://github.com/confidential-containers/cloud-api-adaptor/tree/4111acc80cda3a332927a8b34c009d887c7cbae5/peerpod-ctrl/api/v1alpha1) from peerpod-ctrl to cloud-api-adaptor.
Then the dep will looks like:
```
1. cloud-api-adaptor
2. peerpod-ctrl → cloud-api-adaptor
3. csi-wrapper → cloud-api-adaptor
```

After we break the cycling dep, I agree we move peerpod-ctrl peerpodconfig-ctrl to separate repo.
If we choose option 1, then we will have three new repos:
- github.com/confidential-containers/peerpod-ctrl 
- github.com/confidential-containers/peerpodconfig-ctrl 
- github.com/confidential-containers/cloud(peerpod-providers maybe is better?)

**Comment by liudalibj** [comment]
> New repos are fine as it will help us with dev and release cadence imho.

I will try to prepare fix codes base on new repos.

**Comment by stevenhorsman** [comment]
@liudalibj - in case it ins useful JT created an draft PR to try and help with this: https://github.com/confidential-containers/cloud-api-adaptor/pull/1583

**Comment by stevenhorsman** [comment]
@liudalibj - Do you think this can be closed now that #1719 has merged?

**Comment by liudalibj** [comment]
Yeah, we can close this.
